File: dart/leet_code/950.dart
/// Dart implementation of the LeetCode problem 950: Reveal Cards In Increasing Order.
///
/// This module contains a class `Solution` with a method `deckRevealedIncreasing`.
/// The `deckRevealedIncreasing` method takes in one parameter:
/// - `deck`: A list of integers representing the deck of cards.
///
/// The method returns a list of integers representing the order of revealing the cards.
/// The deck is revealed in the following way: take the top card of the deck, show it, and put it on the table.
/// Then, take the next card of the deck, put it at the bottom of the deck, and continue until all cards are revealed.
///
/// Error Handling:
/// - This method assumes that the input parameter is well-formed, i.e., `deck` is a list of integers.
/// - If the input parameter is not well-formed, the behavior of the method is undefined.

import 'dart:collection';

class Solution {
  /// Reveals the cards in increasing order.
  ///
  /// This method uses a queue to simulate the process of revealing the cards.
  /// It sorts the deck in ascending order and then iterates over the deck from the last card to the first card.
  /// For each card, if the queue is not empty, it moves the first card in the queue to the end of the queue.
  /// Then, it adds the current card to the queue.
  /// Finally, it returns the order of revealing the cards.
  ///
  /// @param deck A list of integers representing the deck of cards.
  /// @return A list of integers representing the order of revealing the cards.
  List<int> deckRevealedIncreasing(List<int> deck) {
    deck.sort();
    Queue<int> dq = Queue<int>();
    for (int i = deck.length - 1; i >= 0; i--) {
      if (dq.isNotEmpty) {
        dq.add(dq.removeFirst());
      }
      dq.add(deck[i]);
    }
    return dq.toList();
  }
}

File: rust/src/leet_code/950.rs
use std::collections::VecDeque;

/// `Solution` is a struct that will contain the methods for solving the problem.
struct Solution;

impl Solution {
    /// This function takes a vector of integers as input, sorts it in increasing order,
    /// and then rearranges it in a specific way: it simulates a process of revealing cards
    /// by taking the top card of the deck, revealing it, and then putting the next card to the bottom of the deck.
    ///
    /// # Arguments
    ///
    /// * `deck` - A vector of integers representing the deck of cards.
    ///
    /// # Returns
    ///
    /// * A vector of integers representing the final state of the deck after the revealing process.
    pub fn deck_revealed_increasing(deck: Vec<i32>) -> Vec<i32> {
        // Create a mutable copy of the deck and sort it
        let mut deck = deck;
        deck.sort();

        // Get the number of cards in the deck
        let n = deck.len();

        // Create a queue to simulate the process of revealing cards and putting the next card to the bottom
        let mut q: VecDeque<usize> = (0..n).collect();

        // Initialize the answer vector with zeros
        let mut ans = vec![0; n];

        // Iterate over the sorted deck
        for &card in deck.iter() {
            // Pop the top card from the queue and reveal it
            let i = q.pop_front().unwrap();
            ans[i] = card;

            // If there are still cards in the queue, put the next card to the bottom
            if let Some(j) = q.pop_front() {
                q.push_back(j);
            }
        }

        // Return the final state of the deck
        ans
    }
}
